File: bindings/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
ABI Framework Basic Usage Example

Demonstrates core functionality including initialization,
vector operations, and database usage.
"""

import sys
import os

# Add parent directory to path for development
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

import abi


def main():
    print("=" * 60)
    print("ABI Framework Basic Usage Example")
    print("=" * 60)

    # Initialize the framework
    print("\n1. Initializing ABI framework...")
    abi.init()
    print(f"   Version: {abi.version()}")
    print(f"   Initialized: {abi.is_initialized()}")
    print(f"   SIMD available: {abi.has_simd()}")

    # Vector operations
    print("\n2. Vector Operations")
    print("-" * 40)

    a = [1.0, 2.0, 3.0, 4.0]
    b = [4.0, 3.0, 2.0, 1.0]

    similarity = abi.cosine_similarity(a, b)
    print(f"   Vectors: a = {a}")
    print(f"            b = {b}")
    print(f"   Cosine similarity: {similarity:.4f}")

    dot = abi.vector_dot(a, b)
    print(f"   Dot product: {dot:.4f}")

    sum_vec = abi.vector_add(a, b)
    print(f"   Sum: {sum_vec}")

    norm = abi.l2_norm(a)
    print(f"   L2 norm of a: {norm:.4f}")

    # Vector database
    print("\n3. Vector Database")
    print("-" * 40)

    db = abi.VectorDatabase(name="example_db", dimensions=4)
    print(f"   Created database: {db.name}")

    # Add vectors with metadata
    db.add([1.0, 0.0, 0.0, 0.0], metadata={"label": "x-axis", "category": "axis"})
    db.add([0.0, 1.0, 0.0, 0.0], metadata={"label": "y-axis", "category": "axis"})
    db.add([0.0, 0.0, 1.0, 0.0], metadata={"label": "z-axis", "category": "axis"})
    db.add([0.0, 0.0, 0.0, 1.0], metadata={"label": "w-axis", "category": "axis"})
    db.add([0.5, 0.5, 0.0, 0.0], metadata={"label": "diagonal-xy", "category": "diagonal"})

    print(f"   Added {db.count} vectors")

    # Search for similar vectors
    query = [0.9, 0.1, 0.0, 0.0]
    results = db.search(query, top_k=3)
    print(f"   Query: {query}")
    print(f"   Top 3 results:")
    for r in results:
        print(f"      ID: {r.id}, Score: {r.score:.4f}, Label: {r.metadata.get('label')}")

    # Get statistics
    stats = db.stats()
    print(f"   Database stats:")
    print(f"      Vectors: {stats.vector_count}")
    print(f"      Dimensions: {stats.dimensions}")
    print(f"      Index type: {stats.index_type}")

    # Agent example
    print("\n4. AI Agent")
    print("-" * 40)

    agent = abi.Agent(name="assistant")
    response = agent.process("Hello! What can you do?")
    print(f"   User: Hello! What can you do?")
    print(f"   Agent: {response}")

    response = agent.process("Tell me about vectors.")
    print(f"   User: Tell me about vectors.")
    print(f"   Agent: {response}")

    print(f"   Conversation history: {len(agent.get_history())} messages")

    # Cleanup
    print("\n5. Cleanup")
    print("-" * 40)
    abi.shutdown()
    print("   Framework shut down successfully")

    print("\n" + "=" * 60)
    print("Example completed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: bindings/wasm/examples/basic-usage.ts
/**
 * Basic usage example for @anthropic/abi-wasm
 *
 * This example demonstrates the core functionality of the ABI WASM bindings.
 *
 * Run with:
 *   npx tsx examples/basic-usage.ts
 */

import {
  init,
  shutdown,
  version,
  isReady,
  cosineSimilarity,
  dotProduct,
  l2Norm,
  normalize,
  VectorDatabase,
} from "../src/index";

async function main() {
  console.log("=== ABI WASM Basic Usage Example ===\n");

  // Initialize the WASM module
  console.log("Initializing ABI WASM...");
  await init({
    // For local development, point to the built WASM file
    wasmPath: "../../zig-out/wasm/abi.wasm",
  });

  console.log(`Initialized: ${isReady()}`);
  console.log(`Version: ${version()}\n`);

  // Vector operations
  console.log("=== Vector Operations ===\n");

  const vectorA = [1.0, 2.0, 3.0, 4.0];
  const vectorB = [4.0, 3.0, 2.0, 1.0];

  console.log(`Vector A: [${vectorA.join(", ")}]`);
  console.log(`Vector B: [${vectorB.join(", ")}]`);
  console.log(`Cosine Similarity: ${cosineSimilarity(vectorA, vectorB).toFixed(4)}`);
  console.log(`Dot Product: ${dotProduct(vectorA, vectorB)}`);
  console.log(`L2 Norm of A: ${l2Norm(vectorA).toFixed(4)}`);
  console.log(`Normalized A: [${normalize(vectorA).map((v) => v.toFixed(4)).join(", ")}]\n`);

  // Vector database
  console.log("=== Vector Database ===\n");

  const db = new VectorDatabase({ name: "embeddings" });

  // Add some vectors representing 3D axes
  const idX = db.add([1.0, 0.0, 0.0], { label: "x-axis", color: "red" });
  const idY = db.add([0.0, 1.0, 0.0], { label: "y-axis", color: "green" });
  const idZ = db.add([0.0, 0.0, 1.0], { label: "z-axis", color: "blue" });

  // Add a diagonal vector
  const idDiag = db.add(normalize([1.0, 1.0, 1.0]), { label: "diagonal" });

  console.log(`Database: ${db.getName()}`);
  console.log(`Size: ${db.size()} vectors`);
  console.log(`Dimensions: ${db.getDimensions()}\n`);

  // Search for vectors similar to a query
  const query = [0.9, 0.1, 0.0]; // Mostly pointing towards X
  console.log(`Query: [${query.join(", ")}]`);
  console.log("Search results (top 3):");

  const results = db.search(query, 3);
  for (const result of results) {
    console.log(
      `  ID: ${result.id}, Score: ${result.score.toFixed(4)}, ` +
        `Label: ${(result.metadata as Record<string, string>)?.label ?? "N/A"}`
    );
  }

  console.log("\n=== Done ===");

  // Cleanup
  shutdown();
}

main().catch(console.error);
